scripts/py_matter_yamltests/matter_yamltests/constraints.py

```diff
@@ -18,57 +18,213 @@
+
+
+class ConstraintError(ConstraintCheckError):
+    def __init__(self, context, reason):
+        super().__init__(context, '', reason)
+
+
+class ConstraintHasValueError(ConstraintCheckError):
+    def __init__(self, context, reason):
+        super().__init__(context, 'hasValue', reason)
+
+
+class ConstraintMinLengthError(ConstraintCheckError):
+    def __init__(self, context, reason):
+        super().__init__(context, 'minLength', reason)
+
+
+class ConstraintMaxLengthError(ConstraintCheckError):
+    def __init__(self, context, reason):
+        super().__init__(context, 'maxLength', reason)
+
+
+class ConstraintIsHexStringError(ConstraintCheckError):
+    def __init__(self, context, reason):
+        super().__init__(context, 'isHexString', reason)
+
+
+class ConstraintStartsWithError(ConstraintCheckError):
+    def __init__(self, context, reason):
+        super().__init__(context, 'startsWith', reason)
+
+
+class ConstraintEndsWithError(ConstraintCheckError):
+    def __init__(self, context, reason):
+        super().__init__(context, 'endsWith', reason)
+
+
+class ConstraintIsUpperCaseError(ConstraintCheckError):
+    def __init__(self, context, reason):
+        super().__init__(context, 'isUpperCase', reason)
+
+
+class ConstraintIsLowerCaseError(ConstraintCheckError):
+    def __init__(self, context, reason):
+        super().__init__(context, 'isLowerCase', reason)
+
+
+class ConstraintMinValueError(ConstraintCheckError):
+    def __init__(self, context, reason):
+        super().__init__(context, 'minValue', reason)
+
+
+class ConstraintMaxValueError(ConstraintCheckError):
+    def __init__(self, context, reason):
+        super().__init__(context, 'maxValue', reason)
+
+
+class ConstraintNotValueError(ConstraintCheckError):
+    def __init__(self, context, reason):
+        super().__init__(context, 'notValue', reason)
+
+
 class BaseConstraint(ABC):
     '''Constraint Interface'''
 
-    def __init__(self, types: list, is_null_allowed: bool = False):
+    def __init__(self, context, types: list, is_null_allowed: bool = False):
         '''An empty type list provided that indicates any type is accepted'''
         self._types = types
         self._is_null_allowed = is_null_allowed
+        self._context = context
 
     def is_met(self, value, value_type_name):
-        if value is None:
-            return self._is_null_allowed
+        if value is None and self._is_null_allowed:
+            return True
 
         response_type = type(value)
         if self._types:
             found_type_match = any(
                 [issubclass(response_type, expected) for expected in self._types])
             if not found_type_match:
-                return False
+                if len(self._types) == 1:
+                    expected_str = f'type "{self._types[0].__name__}"'
+                else:
+                    expected_str = f'one of those types: {[x.__name__ for x in self._types]}'
```

Comment:
Will this put commas between the types automatically?

Reply:
Yes. Here is how it looks:
<img width="1783" alt="Screenshot 2023-02-16 at 10 23 02" src="https://user-images.githubusercontent.com/65897387/219323117-19db82cd-408f-466a-a035-789bdca2e034.png">